File: tests/inter-language-test/src/main/cpp/IltConsumerGetterSetterTest.cpp
/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
#include "IltAbstractConsumerTest.h"

using namespace ::testing;

bool IltUtil::useRestricted64BitRange = true;
bool IltUtil::useRestrictedUnsignedRange = true;

// joynr::Logger logger("IltAbstractConsumerTest");

class IltConsumerGetterSetterTest : public IltAbstractConsumerTest<::testing::Test>
{
public:
    IltConsumerGetterSetterTest() = default;

    template <typename AttributeType>
    void genericSetGetTestMethod(
            std::function<void(joynr::interlanguagetest::TestInterfaceSyncProxy*,
                               AttributeType&,
                               boost::optional<joynr::MessagingQos>)> getter,
            std::function<void(joynr::interlanguagetest::TestInterfaceSyncProxy*,
                               const AttributeType,
                               boost::optional<joynr::MessagingQos>)> setter,
            AttributeType expectedResult);
};

template <typename AttributeType>
void IltConsumerGetterSetterTest::genericSetGetTestMethod(
        std::function<void(joynr::interlanguagetest::TestInterfaceSyncProxy*,
                           AttributeType&,
                           boost::optional<joynr::MessagingQos>)> getter,
        std::function<void(joynr::interlanguagetest::TestInterfaceSyncProxy*,
                           const AttributeType,
                           boost::optional<joynr::MessagingQos>)> setter,
        AttributeType expectedResult)
{
    AttributeType result;
    JOYNR_ASSERT_NO_THROW(setter(testInterfaceProxy.get(), expectedResult, {}));
    JOYNR_ASSERT_NO_THROW(getter(testInterfaceProxy.get(), result, {}));
    ASSERT_EQ(result, expectedResult);
}

TEST_F(IltConsumerGetterSetterTest, callSetAttributeUInt8)
{
    JOYNR_ASSERT_NO_THROW(testInterfaceProxy->setAttributeUInt8(127));
}

TEST_F(IltConsumerGetterSetterTest, callGetAttributeUInt8)
{
    genericSetGetTestMethod<std::uint8_t>(
            &joynr::interlanguagetest::TestInterfaceProxy::getAttributeUInt8,
            &joynr::interlanguagetest::TestInterfaceProxy::setAttributeUInt8,
            127);
}

TEST_F(IltConsumerGetterSetterTest, callSetAttributeDouble)
{
    JOYNR_ASSERT_NO_THROW(testInterfaceProxy->setAttributeDouble(1.1));
}

TEST_F(IltConsumerGetterSetterTest, callGetAttributeDouble)
{
    double result = 0;
    double expectedResult = 1.1;
    JOYNR_ASSERT_NO_THROW({
        testInterfaceProxy->setAttributeDouble(expectedResult);
        testInterfaceProxy->getAttributeDouble(result);
    });
    ASSERT_TRUE(IltUtil::cmpDouble(result, expectedResult));
}

TEST_F(IltConsumerGetterSetterTest, callGetAttributeBooleanReadonly)
{
    bool result;
    JOYNR_ASSERT_NO_THROW({ testInterfaceProxy->getAttributeBooleanReadonly(result); });
    ASSERT_TRUE(result);
}

TEST_F(IltConsumerGetterSetterTest, callSetAttributeStringNoSubscriptions)
{
    JOYNR_ASSERT_NO_THROW(testInterfaceProxy->setAttributeStringNoSubscriptions("Hello world"));
}

TEST_F(IltConsumerGetterSetterTest, callGetAttributeStringNoSubscriptions)
{
    genericSetGetTestMethod<std::string>(
            &joynr::interlanguagetest::TestInterfaceProxy::getAttributeStringNoSubscriptions,
            &joynr::interlanguagetest::TestInterfaceProxy::setAttributeStringNoSubscriptions,
            "Hello world");
}

TEST_F(IltConsumerGetterSetterTest, callGetAttributeInt8readonlyNoSubscriptions)
{
    int8_t result;
    int8_t expectedResult = -128;
    JOYNR_ASSERT_NO_THROW({ testInterfaceProxy->getAttributeInt8readonlyNoSubscriptions(result); });
    ASSERT_EQ(result, expectedResult);
}

TEST_F(IltConsumerGetterSetterTest, callGetAttributeByteBuffer)
{
    genericSetGetTestMethod<joynr::ByteBuffer>(
            &joynr::interlanguagetest::TestInterfaceProxy::getAttributeByteBuffer,
            &joynr::interlanguagetest::TestInterfaceProxy::setAttributeByteBuffer,
            joynr::ByteBuffer{0, 100, 255});
}

TEST_F(IltConsumerGetterSetterTest, callGetAttributeInt64TypeDef)
{
    genericSetGetTestMethod<std::int64_t>(
            &joynr::interlanguagetest::TestInterfaceProxy::getAttributeInt64TypeDef,
            &joynr::interlanguagetest::TestInterfaceProxy::setAttributeInt64TypeDef,
            1L);
}

TEST_F(IltConsumerGetterSetterTest, callGetAttributeStringTypeDef)
{
    genericSetGetTestMethod<joynr::interlanguagetest::typeDefCollection::TypeDefForString>(
            &joynr::interlanguagetest::TestInterfaceProxy::getAttributeStringTypeDef,
            &joynr::interlanguagetest::TestInterfaceProxy::setAttributeStringTypeDef,
            "TypeDefTestString");
}

TEST_F(IltConsumerGetterSetterTest, callGetAttributeStructTypeDef)
{
    genericSetGetTestMethod<joynr::interlanguagetest::typeDefCollection::TypeDefForStruct>(
            &joynr::interlanguagetest::TestInterfaceProxy::getAttributeStructTypeDef,
            &joynr::interlanguagetest::TestInterfaceProxy::setAttributeStructTypeDef,
            IltUtil::createBaseStruct());
}

TEST_F(IltConsumerGetterSetterTest, callGetAttributeMapTypeDef)
{
    using testType = joynr::interlanguagetest::typeDefCollection::TypeDefForMap;
    testType expectedResult;
    expectedResult.insert(std::pair<std::string, std::string>("keyString1", "valueString1"));
    expectedResult.insert(std::pair<std::string, std::string>("keyString2", "valueString2"));
    expectedResult.insert(std::pair<std::string, std::string>("keyString3", "valueString3"));
    genericSetGetTestMethod<testType>(
            &joynr::interlanguagetest::TestInterfaceProxy::getAttributeMapTypeDef,
            &joynr::interlanguagetest::TestInterfaceProxy::setAttributeMapTypeDef,
            expectedResult);
}

TEST_F(IltConsumerGetterSetterTest, callGetAttributeEnumTypeDef)
{
    using testType = joynr::interlanguagetest::typeDefCollection::TypeDefForEnum::Enum;
    testType expectedResult =
            joynr::interlanguagetest::typeDefCollection::TypeDefForEnum::ENUM_0_VALUE_1;
    genericSetGetTestMethod<testType>(
            &joynr::interlanguagetest::TestInterfaceProxy::getAttributeEnumTypeDef,
            &joynr::interlanguagetest::TestInterfaceProxy::setAttributeEnumTypeDef,
            expectedResult);
}

TEST_F(IltConsumerGetterSetterTest, callGetAttributeByteBufferTypeDef)
{
    using testType = joynr::interlanguagetest::typeDefCollection::TypeDefForByteBuffer;
    testType expectedResult = {0x00, 0x64, 0xFF};
    genericSetGetTestMethod<testType>(
            &joynr::interlanguagetest::TestInterfaceProxy::getAttributeByteBufferTypeDef,
            &joynr::interlanguagetest::TestInterfaceProxy::setAttributeByteBufferTypeDef,
            expectedResult);
}

TEST_F(IltConsumerGetterSetterTest, callGetAttributeArrayTypeDef)
{
    using testType = joynr::interlanguagetest::typeDefCollection::ArrayTypeDefStruct;
    std::vector<std::string> stringArray = IltUtil::createStringArray();
    testType expectedResult;
    expectedResult.setTypeDefStringArray(stringArray);
    genericSetGetTestMethod<testType>(
            &joynr::interlanguagetest::TestInterfaceProxy::getAttributeArrayTypeDef,
            &joynr::interlanguagetest::TestInterfaceProxy::setAttributeArrayTypeDef,
            expectedResult);
}

TEST_F(IltConsumerGetterSetterTest, callSetAttributeArrayOfStringImplicit)
{
    std::vector<std::string> stringArrayArg = IltUtil::createStringArray();
    JOYNR_ASSERT_NO_THROW(
            { testInterfaceProxy->setAttributeArrayOfStringImplicit(stringArrayArg); });
}

TEST_F(IltConsumerGetterSetterTest, callGetAttributeArrayOfStringImplicit)
{
    std::vector<std::string> result;
    std::vector<std::string> stringArrayArg = IltUtil::createStringArray();
    JOYNR_ASSERT_NO_THROW({
        testInterfaceProxy->setAttributeArrayOfStringImplicit(stringArrayArg);
        testInterfaceProxy->getAttributeArrayOfStringImplicit(result);
    });
    ASSERT_TRUE(IltUtil::checkStringArray(result));
}

TEST_F(IltConsumerGetterSetterTest, callSetAttributeEnumeration)
{
    JOYNR_ASSERT_NO_THROW({
        joynr::interlanguagetest::Enumeration::Enum enumerationArg =
                joynr::interlanguagetest::Enumeration::ENUM_0_VALUE_2;
        testInterfaceProxy->setAttributeEnumeration(enumerationArg);
    });
}

TEST_F(IltConsumerGetterSetterTest, callGetAttributeEnumeration)
{
    genericSetGetTestMethod<joynr::interlanguagetest::Enumeration::Enum>(
            &joynr::interlanguagetest::TestInterfaceProxy::getAttributeEnumeration,
            &joynr::interlanguagetest::TestInterfaceProxy::setAttributeEnumeration,
            joynr::interlanguagetest::Enumeration::ENUM_0_VALUE_2);
}

TEST_F(IltConsumerGetterSetterTest, callGetAttributeExtendedEnumerationReadonly)
{
    joynr::interlanguagetest::namedTypeCollection2::ExtendedEnumerationWithPartlyDefinedValues::Enum
            result;
    JOYNR_ASSERT_NO_THROW({ testInterfaceProxy->getAttributeExtendedEnumerationReadonly(result); });
    ASSERT_EQ(result,
              joynr::interlanguagetest::namedTypeCollection2::
                      ExtendedEnumerationWithPartlyDefinedValues::
                              ENUM_2_VALUE_EXTENSION_FOR_ENUM_WITHOUT_DEFINED_VALUES);
}

TEST_F(IltConsumerGetterSetterTest, callSetAttributeBaseStruct)
{
    joynr::interlanguagetest::namedTypeCollection2::BaseStruct baseStructArg =
            IltUtil::createBaseStruct();
    JOYNR_ASSERT_NO_THROW({ testInterfaceProxy->setAttributeBaseStruct(baseStructArg); });
}

TEST_F(IltConsumerGetterSetterTest, callGetAttributeBaseStruct)
{
    joynr::interlanguagetest::namedTypeCollection2::BaseStruct result;
    joynr::interlanguagetest::namedTypeCollection2::BaseStruct baseStructArg =
            IltUtil::createBaseStruct();
    JOYNR_ASSERT_NO_THROW({
        testInterfaceProxy->setAttributeBaseStruct(baseStructArg);
        testInterfaceProxy->getAttributeBaseStruct(result);
    });
    ASSERT_TRUE(IltUtil::checkBaseStruct(result));
}

TEST_F(IltConsumerGetterSetterTest, callSetAttributeExtendedExtendedBaseStruct)
{
    joynr::interlanguagetest::namedTypeCollection2::ExtendedExtendedBaseStruct baseStructArg =
            IltUtil::createExtendedExtendedBaseStruct();
    JOYNR_ASSERT_NO_THROW(
            { testInterfaceProxy->setAttributeExtendedExtendedBaseStruct(baseStructArg); });
}

TEST_F(IltConsumerGetterSetterTest, callGetAttributeExtendedExtendedBaseStruct)
{
    joynr::interlanguagetest::namedTypeCollection2::ExtendedExtendedBaseStruct result;
    joynr::interlanguagetest::namedTypeCollection2::ExtendedExtendedBaseStruct baseStructArg =
            IltUtil::createExtendedExtendedBaseStruct();
    JOYNR_ASSERT_NO_THROW({
        testInterfaceProxy->setAttributeExtendedExtendedBaseStruct(baseStructArg);
        testInterfaceProxy->getAttributeExtendedExtendedBaseStruct(result);
    });
    ASSERT_TRUE(IltUtil::checkExtendedExtendedBaseStruct(result));
}

TEST_F(IltConsumerGetterSetterTest, callSetAttributeMapStringString)
{
    joynr::interlanguagetest::namedTypeCollection2::MapStringString mapStringStringArg;
    mapStringStringArg.insert(std::pair<std::string, std::string>("keyString1", "valueString1"));
    mapStringStringArg.insert(std::pair<std::string, std::string>("keyString2", "valueString2"));
    mapStringStringArg.insert(std::pair<std::string, std::string>("keyString3", "valueString3"));
    JOYNR_ASSERT_NO_THROW({ testInterfaceProxy->setAttributeMapStringString(mapStringStringArg); });
}

TEST_F(IltConsumerGetterSetterTest, callGetAttributeMapStringString)
{
    using testType = joynr::interlanguagetest::namedTypeCollection2::MapStringString;
    testType expectedResult;
    expectedResult.insert(std::pair<std::string, std::string>("keyString1", "valueString1"));
    expectedResult.insert(std::pair<std::string, std::string>("keyString2", "valueString2"));
    expectedResult.insert(std::pair<std::string, std::string>("keyString3", "valueString3"));

    genericSetGetTestMethod<testType>(
            &joynr::interlanguagetest::TestInterfaceProxy::getAttributeMapStringString,
            &joynr::interlanguagetest::TestInterfaceProxy::setAttributeMapStringString,
            expectedResult);
}

TEST_F(IltConsumerGetterSetterTest, callGetAttributeMapTypeDefStringString)
{
    using testType = joynr::interlanguagetest::typeDefCollection::TypeDefForMap;
    testType expectedResult;
    expectedResult.insert(std::pair<std::string, std::string>("keyString1", "valueString1"));
    expectedResult.insert(std::pair<std::string, std::string>("keyString2", "valueString2"));
    expectedResult.insert(std::pair<std::string, std::string>("keyString3", "valueString3"));

    genericSetGetTestMethod<testType>(
            &joynr::interlanguagetest::TestInterfaceProxy::getAttributeMapTypeDef,
            &joynr::interlanguagetest::TestInterfaceProxy::setAttributeMapTypeDef,
            expectedResult);
}

TEST_F(IltConsumerGetterSetterTest, callSetAttributeWithExceptionFromSetter)
{
    try {
        testInterfaceProxy->setAttributeWithExceptionFromSetter(false);
        FAIL() << "callSetAttributeWithExceptionFromSetter - Unexpected continuation without "
                  "exception";
    } catch (joynr::exceptions::ProviderRuntimeException& e) {
        if (e.getMessage() != "Exception from setAttributeWithExceptionFromSetter") {
            FAIL() << "callSetAttributeWithExceptionFromSetter - invalid exception message";
        }
        // OK
    } catch (joynr::exceptions::JoynrRuntimeException& e) {
        FAIL() << "callSetAttributeWithExceptionFromSetter - unexpected exception type";
    } catch (...) {
        FAIL() << "callSetAttributeWithExceptionFromSetter: unknown exception caught";
    }
}

TEST_F(IltConsumerGetterSetterTest, callGetAttributeWithExceptionFromGetter)
{
    try {
        bool result;
        testInterfaceProxy->getAttributeWithExceptionFromGetter(result);
        FAIL() << "callGetAttributeWithExceptionFromGetter - Unexpected continuation without "
                  "exception";
    } catch (joynr::exceptions::ProviderRuntimeException& e) {
        if (e.getMessage() != "Exception from getAttributeWithExceptionFromGetter") {
            FAIL() << "callGetAttributeWithExceptionFromGetter - invalid exception message";
        }
        // OK
    } catch (joynr::exceptions::JoynrRuntimeException& e) {
        FAIL() << "callGetAttributeWithExceptionFromGetter - unexpected exception type";
    } catch (...) {
        FAIL() << "callGetAttributeWithExceptionFromGetter: unknown exception caught";
    }
}

File: tests/inter-language-test/inter-language-test-jee-consumer/src/main/java/io/joynr/test/interlanguage/jee/IltConsumerGetterSetterTest.java
/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package io.joynr.test.interlanguage.jee;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertSame;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

import java.lang.reflect.Method;
import java.util.Objects;

import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import joynr.exceptions.ProviderRuntimeException;
import joynr.interlanguagetest.Enumeration;
import joynr.interlanguagetest.namedTypeCollection2.BaseStruct;
import joynr.interlanguagetest.namedTypeCollection2.ExtendedEnumerationWithPartlyDefinedValues;
import joynr.interlanguagetest.namedTypeCollection2.ExtendedExtendedBaseStruct;
import joynr.interlanguagetest.namedTypeCollection2.MapStringString;
import joynr.interlanguagetest.typeDefCollection.ArrayTypeDefStruct;

public class IltConsumerGetterSetterTest extends IltConsumerTest {
    private static final Logger logger = LoggerFactory.getLogger(IltConsumerGetterSetterTest.class);

    @SuppressWarnings("unchecked")
    private <T> void genericGetterSetterTestMethod(T arg, String methodName) {
        try {
            Method mSetter = testInterfaceProxy.getClass().getMethod("set" + methodName, arg.getClass());
            Method mGetter = testInterfaceProxy.getClass().getMethod("get" + methodName);

            mSetter.invoke(testInterfaceProxy, arg);
            T result = (T) mGetter.invoke(testInterfaceProxy);

            assertNotNull(name.getMethodName() + " - FAILED - got no result", result);
            assertTrue(name.getMethodName() + " - FAILED - got invalid result", Objects.deepEquals(arg, result));

        } catch (Exception e) {
            fail(name.getMethodName() + " - FAILED - caught unexpected exception: " + e.getMessage());
        }
    }

    /*
     * GETTER AND SETTER CALLS
     */

    @Test
    public void callSetAndGetAttributeUInt8() {
        logger.info(name.getMethodName());
        byte byteArg = 127;
        genericGetterSetterTestMethod(byteArg, "AttributeUInt8");
        logger.info(name.getMethodName() + " - OK");
    }

    @Test
    public void callSetAndGetAttributeDouble() {
        logger.info(name.getMethodName());
        double doubleArg = 1.1d;
        genericGetterSetterTestMethod(doubleArg, "AttributeDouble");
        logger.info(name.getMethodName() + " - OK");
    }

    @Test
    public void callGetAttributeBooleanReadOnly() {
        logger.info(name.getMethodName());
        try {
            // value is hardcoded on provider side since it is readonly
            Boolean result = testInterfaceProxy.getAttributeBooleanReadonly();
            assertNotNull(name.getMethodName() + " - FAILED - got no result", result);
            assertTrue(name.getMethodName() + " - FAILED - got invalid result", result);
        } catch (Exception e) {
            fail(name.getMethodName() + " - FAILED - caught unexpected exception: " + e.getMessage());
        }

        logger.info(name.getMethodName() + " - OK");
    }

    // there is no setter for attributeBooleanReadOnly, but this cannot
    // be checked since it would create a compiler error

    @Test
    public void callSetAndGetAttributeStringNoSubscriptions() {
        logger.info(name.getMethodName());
        String stringArg = "Hello world";
        genericGetterSetterTestMethod(stringArg, "AttributeStringNoSubscriptions");
        logger.info(name.getMethodName() + " - OK");
    }

    @Test
    public void callSetAndGetAttributeInt8readonlyNoSubscriptions() {
        logger.info(name.getMethodName());
        try {
            // value is hardcoded on provider side since it is readonly
            Byte result = testInterfaceProxy.getAttributeInt8readonlyNoSubscriptions();
            assertNotNull(name.getMethodName() + " - FAILED - got no result", result);
            assertEquals(name.getMethodName() + " - FAILED - got invalid result", -128, (byte) result);
        } catch (Exception e) {
            fail(name.getMethodName() + " - FAILED - caught unexpected exception: " + e.getMessage());
        }

        logger.info(name.getMethodName() + " - OK");
    }

    // no setter for attributeInt8readonlyNoSubscriptions

    @Test
    public void callSetAndGetAttributeArrayOfStringImplicit() {
        logger.info(name.getMethodName());
        String[] stringArrayArg = { "Hello", "World" };
        genericGetterSetterTestMethod(stringArrayArg, "AttributeArrayOfStringImplicit");
        logger.info(name.getMethodName() + " - OK");
    }

    @Test
    public void callSetAndSetAndGetAttributeByteBuffer() {
        logger.info(name.getMethodName());
        Byte[] byteBufferArg = { -128, 0, 127 };
        genericGetterSetterTestMethod(byteBufferArg, "AttributeByteBuffer");
        logger.info(name.getMethodName() + " - OK");
    }

    @Test
    public void callSetAndGetAttributeInt64TypeDef() {
        logger.info(name.getMethodName());
        Long int64TypeDefArg = 1L;
        genericGetterSetterTestMethod(int64TypeDefArg, "AttributeInt64TypeDef");
        logger.info(name.getMethodName() + " - OK");
    }

    @Test
    public void callSetAndGetAttributeStringTypeDef() {
        logger.info(name.getMethodName());
        String stringTypeDefArg = "StringTypeDef";
        genericGetterSetterTestMethod(stringTypeDefArg, "AttributeStringTypeDef");
        logger.info(name.getMethodName() + " - OK");
    }

    @Test
    public void callSetAndGetAttributeStructTypeDef() {
        logger.info(name.getMethodName());
        BaseStruct structTypeDefArg = IltUtil.createBaseStruct();
        genericGetterSetterTestMethod(structTypeDefArg, "AttributeStructTypeDef");
        logger.info(name.getMethodName() + " - OK");
    }

    @Test
    public void callSetAndGetGetAttributeMapTypeDef() {
        logger.info(name.getMethodName());
        MapStringString mapTypeDefArg = new MapStringString();
        mapTypeDefArg.put("keyString1", "valueString1");
        mapTypeDefArg.put("keyString2", "valueString2");
        mapTypeDefArg.put("keyString3", "valueString3");
        genericGetterSetterTestMethod(mapTypeDefArg, "AttributeMapTypeDef");
        logger.info(name.getMethodName() + " - OK");
    }

    @Test
    public void callSetAndGetGetAttributeEnumTypeDef() {
        logger.info(name.getMethodName());
        Enumeration enumTypeDefArg = Enumeration.ENUM_0_VALUE_1;
        genericGetterSetterTestMethod(enumTypeDefArg, "AttributeEnumTypeDef");
        logger.info(name.getMethodName() + " - OK");
    }

    @Test
    public void callSetAndGetGetAttributeByteBufferTypeDef() {
        logger.info(name.getMethodName());
        Byte[] byteBufferTypeDefArg = { -128, 0, 127 };
        genericGetterSetterTestMethod(byteBufferTypeDefArg, "AttributeByteBufferTypeDef");
        logger.info(name.getMethodName() + " - OK");
    }

    @Test
    public void callSetAndGetGetAttributeArrayTypeDef() {
        logger.info(name.getMethodName());
        String[] stringArray = { "Hello", "World" };
        ArrayTypeDefStruct arrayTypeDefArg = new ArrayTypeDefStruct(stringArray);
        genericGetterSetterTestMethod(arrayTypeDefArg, "AttributeArrayTypeDef");
        logger.info(name.getMethodName() + " - OK");
    }

    @Test
    public void callSetAndGetGetAttributeEnumeration() {
        logger.info(name.getMethodName());
        Enumeration enumerationArg = Enumeration.ENUM_0_VALUE_2;
        genericGetterSetterTestMethod(enumerationArg, "AttributeEnumeration");
        logger.info(name.getMethodName() + " - OK");
    }

    @Test
    public void callGetAttributeExtendedEnumerationReadonly() {
        logger.info(name.getMethodName());
        try {
            // value is hardcoded on provider side since it is readonly
            ExtendedEnumerationWithPartlyDefinedValues result = testInterfaceProxy.getAttributeExtendedEnumerationReadonly();
            assertNotNull(name.getMethodName() + " - FAILED - got no result", result);
            assertSame(name.getMethodName() + " - FAILED - got invalid result",
                       result,
                       ExtendedEnumerationWithPartlyDefinedValues.ENUM_2_VALUE_EXTENSION_FOR_ENUM_WITHOUT_DEFINED_VALUES);
        } catch (Exception e) {
            fail(name.getMethodName() + " - FAILED - caught unexpected exception: " + e.getMessage());
        }

        logger.info(name.getMethodName() + " - OK");
    }

    // no setter for attributeExtendedEnumerationReadonly

    @Test
    public void callSetAndGetGetAttributeBaseStruct() {
        logger.info(name.getMethodName());
        BaseStruct baseStructArg = IltUtil.createBaseStruct();
        genericGetterSetterTestMethod(baseStructArg, "AttributeBaseStruct");
        logger.info(name.getMethodName() + " - OK");
    }

    @Test
    public void callSetAndGetAttributeExtendedExtendedBaseStruct() {
        logger.info(name.getMethodName());
        ExtendedExtendedBaseStruct extendedExtendedBaseStructArg = IltUtil.createExtendedExtendedBaseStruct();
        genericGetterSetterTestMethod(extendedExtendedBaseStructArg, "AttributeExtendedExtendedBaseStruct");
        logger.info(name.getMethodName() + " - OK");
    }

    @Test
    public void callSetAndGetAttributeMapStringString() {
        logger.info(name.getMethodName());
        MapStringString attributeMapStringStringArg = new MapStringString();
        attributeMapStringStringArg.put("keyString1", "valueString1");
        attributeMapStringStringArg.put("keyString2", "valueString2");
        attributeMapStringStringArg.put("keyString3", "valueString3");
        genericGetterSetterTestMethod(attributeMapStringStringArg, "AttributeMapStringString");
        logger.info(name.getMethodName() + " - OK");
    }

    /*
     * GETTER AND SETTER CALLS WITH EXCEPTION
     */

    @Test
    public void callGetAttributeWithExceptionFromGetter() {
        logger.info(name.getMethodName());
        try {
            testInterfaceProxy.getAttributeWithExceptionFromGetter();
            fail(name.getMethodName() + " - FAILED - unexpected return without exception");
        } catch (ProviderRuntimeException e) {
            if (e.getMessage() == null
                    || !e.getMessage().endsWith("Exception from getAttributeWithExceptionFromGetter")) {
                fail(name.getMethodName() + " - FAILED - caught invalid exception: " + e.getMessage());
            }
        } catch (Exception e) {
            fail(name.getMethodName() + " - FAILED - caught unexpected exception: " + e.getMessage());
        }

        logger.info(name.getMethodName() + " - OK");
    }

    @Test
    public void callSetAttributeWithExceptionFromSetter() {
        logger.info(name.getMethodName());
        try {
            testInterfaceProxy.setAttributeWithExceptionFromSetter(false);
            fail(name.getMethodName() + " - FAILED - got no result");
        } catch (ProviderRuntimeException e) {
            if (e.getMessage() == null
                    || !e.getMessage().endsWith("Exception from setAttributeWithExceptionFromSetter")) {
                fail(name.getMethodName() + " - FAILED - caught invalid exception: " + e.getMessage());
            }
        } catch (Exception e) {
            fail(name.getMethodName() + " - FAILED - caught unexpected exception: " + e.getMessage());
        }

        logger.info(name.getMethodName() + " - OK");
    }
}
